tools/copy_directory/main.go

```diff
@@ -0,0 +1,202 @@
+package main
+
+import (
+	"fmt"
+	"io"
+	"io/fs"
+	"log"
+	"os"
+	"path"
+	"path/filepath"
+	"strings"
+)
+
+type copyFile struct {
+	Src      string
+	Dst      string
+	FileInfo fs.FileInfo
+}
+
+type copySet []copyFile
+type dirSet []string
+type pathSet map[string]bool
+
+var copyPaths = make(copySet, 0, 10*1024)
+var dstDirs = make(dirSet, 0, 1024)
+var srcPaths = pathSet{}
+
+func scanDir(src string, dst string) error {
+	// filepath.WalkDir walks the file tree rooted at root, calling fn for each file or directory in
+	// the tree, including root. See https://pkg.go.dev/path/filepath#WalkDir for more info.
+	return filepath.WalkDir(src, func(p string, dirEntry fs.DirEntry, err error) error {
+		r, err := filepath.Rel(src, p)
+		if err != nil {
+			return err
+		}
+
+		d := filepath.Join(dst, r)
+
+		if dirEntry.IsDir() {
+			srcPaths[src] = true
+			dstDirs = append(dstDirs, d)
+			return nil
+		} else {
+			info, err := dirEntry.Info()
+			if err != nil {
+				return err
+			}
+
+			if info.Mode()&os.ModeSymlink == os.ModeSymlink {
+				// symlink to directories are intentionally never followed by filepath.Walk to avoid infinite recursion
+				linkPath, err := os.Readlink(p)
+				if err != nil {
+					return err
+				}
+				if !path.IsAbs(linkPath) {
+					linkPath = path.Join(path.Dir(p), linkPath)
+				}
+				if srcPaths[linkPath] {
+					// recursive symlink; silently ignore
+					return nil
+				}
+				stat, err := os.Stat(linkPath)
+				if err != nil {
+					return fmt.Errorf("failed to stat file %s pointed to by symlink %s: %w", linkPath, p, err)
+				}
+				if stat.IsDir() {
+					// symlink points to a directory
+					return scanDir(linkPath, d)
+				} else {
+					// symlink points to a regular file
+					copyPaths = append(copyPaths, copyFile{
+						Src:      linkPath,
+						Dst:      d,
+						FileInfo: stat,
+					})
+					return nil
+				}
+			}
+
+			// a regular file
+			copyPaths = append(copyPaths, copyFile{
+				Src:      p,
+				Dst:      d,
+				FileInfo: info,
+			})
+			return nil
+		}
+	})
+}
+
+// From https://opensource.com/article/18/6/copying-files-go
+func copy(src string, dst string) error {
+	source, err := os.Open(src)
+	if err != nil {
+		return err
+	}
+	defer source.Close()
+
+	destination, err := os.Create(dst)
+	if err != nil {
+		return err
+	}
+	defer destination.Close()
+	_, err = io.Copy(destination, source)
+	return err
+}
+
+func version() string {
+	var versionBuilder strings.Builder
+	if Release != "" && Release != PreStampRelease {
+		versionBuilder.WriteString(Release)
+		if GitStatus != CleanGitStatus {
+			versionBuilder.WriteString(NotCleanVersionSuffix)
+		}
+	} else {
+		versionBuilder.WriteString(NoReleaseVersion)
+	}
+	return versionBuilder.String()
+}
+
+func main() {
+	flags := make([]string, 0, len(os.Args))
+	args := make([]string, 0, len(os.Args))
```

Comment:
nit: since we expect there to be only two args 

```suggestion
	args := make([]string, 0, 2)
```

Reply:
thinking more about this; maybe this shouldn't be an array at all since there's only two args. how about if we force that `os.Args[1]` is always `src` and `os.Args[2]` is `dst` and followed by flags? then we could get rid of this loop

Reply:
that seems problematic especially since I want the `--version` flag to work if that is the first arg. a small array of 2 ints will do the trick

Reply:
on 2nd thought lets go with that. `--version` is a special case where it is checked in the first arg only